esphome/components/mcp4461/mcp4461.cpp

```diff
@@ -0,0 +1,662 @@
+
+void Mcp4461Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up mcp4461 using address (0x%02" PRIX8 ")...", this->address_);
+  auto err = this->write(nullptr, 0);
+  if (err != i2c::ERROR_OK) {
+    this->error_code_ = MCP4461_STATUS_I2C_ERROR;
+    this->mark_failed();
+    return;
+  }
+  this->begin_();
+}
+
+void Mcp4461Component::begin_() {
+  // save WP/WL status
+  this->update_write_protection_status_();
+  for (uint8_t i = 0; i < 8; i++) {
+    if (this->reg_[i].initial_value.has_value()) {
+      uint16_t initial_state;
+      initial_state = static_cast<uint16_t>(*this->reg_[i].initial_value * 256.0f);
+      this->write_wiper_level_(i, initial_state);
+    }
+    if (this->reg_[i].enabled) {
+      this->reg_[i].state = this->read_wiper_level_(i);
+    } else {
+      // only volatile wipers can be set disabled on hw level
+      if (i < 4) {
+        this->reg_[i].state = 0;
+        Mcp4461WiperIdx wiper_idx;
+        wiper_idx = static_cast<Mcp4461WiperIdx>(i);
+        this->disable_wiper_(wiper_idx);
+      }
+    }
+  }
+}
+
+void Mcp4461Component::set_initial_value(Mcp4461WiperIdx wiper, float initial_value) {
+  uint8_t wiper_idx = static_cast<uint8_t>(wiper);
+  this->reg_[wiper_idx].initial_value = initial_value;
+}
+
+void Mcp4461Component::initialize_terminal_disabled(Mcp4461WiperIdx wiper, char terminal) {
+  uint8_t wiper_idx = static_cast<uint8_t>(wiper);
+  switch (terminal) {
+    case 'a':
+      this->reg_[wiper_idx].terminal_a = false;
+      break;
+    case 'b':
+      this->reg_[wiper_idx].terminal_b = false;
+      break;
+    case 'w':
+      this->reg_[wiper_idx].terminal_w = false;
+      break;
+  }
+}
+
+void Mcp4461Component::update_write_protection_status_() {
+  uint8_t status_register_value;
+  status_register_value = this->get_status_register_();
+  this->write_protected_ = static_cast<bool>((status_register_value >> 0) & 0x01);
+  this->reg_[0].wiper_lock_active = static_cast<bool>((status_register_value >> 2) & 0x01);
+  this->reg_[1].wiper_lock_active = static_cast<bool>((status_register_value >> 3) & 0x01);
+  this->reg_[2].wiper_lock_active = static_cast<bool>((status_register_value >> 5) & 0x01);
+  this->reg_[3].wiper_lock_active = static_cast<bool>((status_register_value >> 6) & 0x01);
+}
+
+void Mcp4461Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "mcp4461:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "%s", LOG_STR_ARG(this->get_message_string(this->error_code_)));
+  }
+  // log wiper status
+  for (uint8_t i = 0; i < 8; ++i) {
+    // terminals only valid for volatile wipers 0-3 - enable/disable is terminal hw
+    // so also invalid for nonvolatile. For these, only print current level.
+    // reworked to be a one-line intentionally, as output would not be in order
+    if (i < 4) {
+      ESP_LOGCONFIG(TAG,
+                    "  ├── Volatile wiper [%" PRIu8 "] level: %" PRIu16 ", Status: %s, HW: %s, A: %s, B: %s, W: %s", i,
```

Comment:
You don't _need_ to use the macros for less-than-32-bit int types (we prefer not to and aren't elsewhere).
```suggestion
                    "  ├── Volatile wiper [%u] level: %u, Status: %s, HW: %s, A: %s, B: %s, W: %s", i,
```

Reply:
Thanks for clarifying it! This was actually my fault. I thought it's preferred to use them everywhere due to issue with idf 4 and 5.